rust/src/utils/kmeans.rs

```diff
@@ -187,33 +186,37 @@ impl KMeanMembership {
     async fn to_kmeans(&self) -> Result<KMeans> {
         let dimension = self.dimension;
         let cluster_ids = Arc::new(self.cluster_ids.clone());
-        let data = self.data.clone();
 
         // New centroids for each cluster
         let means = stream::iter(0..self.k)
             .zip(repeat_with(|| {
                 (
-                    data.clone(),
+                    self.data.clone(),
                     cluster_ids.clone(),
                     self.centroids.clone(),
                 )
             }))
             .map(
                 |(cluster, (data, cluster_ids, prev_centroids))| async move {
                     tokio::task::spawn_blocking(move || {
-                        let mut sum = Float32Array::from_iter_values(
-                            (0..dimension).map(|_| 0.0).collect::<Vec<_>>(),
-                        );
+                        let mut sum = vec![0.0; dimension];
+                        let data = data.values();
                         let mut total = 0.0;
+                        // Eager group-by cluster id.
                         for i in 0..cluster_ids.len() {
                             if cluster_ids[i] as usize == cluster {
-                                sum =
-                                    add(&sum, &data.slice(i * dimension, dimension)).unwrap();
+                                // TODO: use simd ADD
+                                for j in 0..dimension {
+                                    sum[j] += data[i * dimension + j];
+                                }
                                 total += 1.0;
                             };
                         }
                         if total > 0.0 {
-                            divide_scalar(&sum, total).unwrap()
+                            let s = Float32Array::from_iter_values(
+                                sum
+                            );
```

Comment:
@eddyxu FYI `from_iter_values()` iterates over the vector and creates a new buffer for the Float32Array. To be even more efficient, we can use `Float32Array::from`, which will take ownership of the underlying buffer that backs the vector and wraps it as an array. Basically zero copy.

https://docs.rs/arrow-array/41.0.0/src/arrow_array/array/primitive_array.rs.html#1058

Reply:
Nice, TIL.